### DISCUSSION TITLE
Add button component to the component library.

### DISCUSSION THREAD

**Discussion by Dhruvi16** [mathesar-foundation/mathesar#351] [Maintainer discussions]
Link to the Figma file - [Button Component](https://www.figma.com/file/xYEhAhzGZCwTKmUMqstt2Y/Components?node-id=107%3A2)

Main concerns -
1. Button styles
2. Was I able to cover every aspect of a button component? If not, please lemme know what all did I miss. And please do comment on the current doc also.
3. Next steps like - converting them into components, creating buttons for every use case, documenting specs, and updating them with the style guide we choose like color, etc.

**Comment by Dhruvi16** [comment]
@ghislaineguerin @kgodey @pavish please do comment.

**Comment by pavish** [comment]
@Dhruvi16 The button component spec looks good to me. I do have the following comments:
* It would be better to split focus, active and hover states.
  - One button can be hovered upon when another is focused. The focus state should also be similar across different kinds of components.
* The padding could be made relative to the size of the font. Fixed padding does not work well for all font sizes.
* Border-radius needs to be specified explicitly, and it also could be made relative to font size.

**Comment by ghislaineguerin** [comment]
@Dhruvi16 
This is a good start. I wanted to provide some guidelines to improve the review of new components first, before moving to specific feedback regarding the component itself.

I think you can explain what the component is and the context in which users might interact with it? For example: "The button component is used to represent actions users can take." While this might sound obvious, I believe that as we increase the complexity of components, it will be necessary to communicate their purpose and intent.

Intent also helps to describe better the different variations, rather than focusing much on style. For example: "The primary button is the main action a user can take and should use the primary color background with maximum contrast text." If we focus much on style, it will be much harder to transition into different applications such as night mode or further customization options. We might then have rules for how we should calculate contrast thresholds or tonal offsets. This is the basis for having an intention-oriented component.

Another critical aspect of defining a component should be describing its limitations or usages that we should not give to the component. Thinking about this upfront will save a lot of time later and help us build these components more robustly. For example, "Don't use buttons to navigate to a different view. If it navigates, it should be a link." or "To remain functional it is advised that buttons contain no more than 20 characters." etc.

Based on this, I think a good way to present a new component should be:
- Component Description
- Usage Guidelines
- Limitations

The review should then focus on whether all component intentions are optimally resolved. For example, "The main button looks less important than the secondary" or "the active button is too similar to the default one." A good question to ask would be:
- Are the selected styles working as intended? If not, please explain what's failing (e.g., ambiguous, not strong enough compared to the secondary action, inconsistent with other similar intent components).
- Is there confusion regarding the usage of this component? Do the provided examples cover all existing use cases?
Does the component not correctly honor the style guide? For example, "Font size is not correct."
- I would also add, if possible, a review of examples provided. Reviewing usage examples would help visualize implementation and make sure we don't introduce unwanted changes in other components by changing global styles.

@kgodey please let me know if there's anything else you'd like to add to this from the product perspective. 
@pavish In the case of relative sizing and other front-end design concerns, where they cannot be represented in Figma, what would the best approach be? Do you prefer visual examples rather than describing their behavior? A combination of both?

**Reply by kgodey** [reply]
@ghislaineguerin I don't have any further requirements from a product perspective. I do agree that emphasizing intention is the most important thing, and we might want to add further guidelines to each variation of buttons. For instance, we might want to say that in any layout component, there should not be more than one primary button so that the user always knows what the main action is (this is just an example, I'm not sure if that's actually how we intend to use primary buttons).

As far as sizing goes, I think we need to decide whether we want to support buttons of any size or offer specified sizes of buttons (like Bootstrap's small, regular, large sizes). If we only support a few button size variations, then we can figure out what the padding and font size for those variations should be. @pavish do you have any opinions on this?

I also wanted to link to CC's design library sections on [spacing](https://www.figma.com/file/l4Mt3dn3Ndtrvrb4aLcwXI/Design-Library?node-id=0%3A190&viewport=-3514%2C-324%2C0.6717029213905334) and [buttons](https://www.figma.com/file/l4Mt3dn3Ndtrvrb4aLcwXI/Design-Library?node-id=10273%3A10&viewport=-3514%2C-324%2C0.6717029213905334) for inspiration. @ghislaineguerin you can see how spacing is represented in Figma there.

**Reply by pavish** [reply]
If the padding is based on a relative unit (preferably rem), users will be able to use the component on their desired size levels.

The CC's design library links by @kgodey seem to detail the spacing requirements, as expected, in both relative and absolute units. Having something similar would be perfect. @ghislaineguerin

**Comment by Dhruvi16** [comment]
On @pavish's point of view on variation. I reckon we can decide upon a couple of variations of button sizes as @kgodey mentioned. I will work on an example and get back to you so we can have a clear idea. 

@ghislaineguerin thank you for breaking it down this beautifully. I will definitely make sure of the intent in the next iteration. 
And yes @kgodey CC's DL is a great example. I will try and take inspiration as much as possible.

**Comment by Dhruvi16** [comment]
@kgodey @ghislaineguerin @pavish Take a look at the version 2 of the document - [Link](https://www.figma.com/file/xYEhAhzGZCwTKmUMqstt2Y/Components?node-id=154%3A2)

**Reply by kgodey** [reply]
Thanks for working on this, @Dhruvi16! It looks much more comprehensive. I'll save my review until after @ghislaineguerin and @pavish have had a chance to look, since they're the primary consumers.

**Reply by ghislaineguerin** [reply]
@Dhruvi16 Thanks for the update. I will write a review and share it during the day tomorrow.

**Comment by ghislaineguerin** [comment]
Thanks, @Dhruvi16  Some questions:

1. I see that some examples use uppercase labels. Typically, and due to the scale of button labels, I'd choose Capitalized labels as I find them easier to recognize.
2. I'm still not sure we have a strong justification for a tertiary button. Do you think it's best to define two types and add a third once we find a suitable scenario?
3. I think the styles for states look good. The only issue I see is that the hover for the primary button has a very high contrast offset compared to the original. Can we try having them closer in contrast yet noticeable enough?
4. Can we include buttons for particular scenarios, such as delete confirmation (danger)?

**Comment by pavish** [comment]
@Dhruvi16 It looks pretty good :)

Could we also document the sizes (both fixed and relative)? I can see we have representations, it would be better to document it as well.

**Comment by kgodey** [comment]
Looks good, @Dhruvi16! My feedback:

- I think it would be clearer to name the buttons "primary", "secondary" etc. rather than "contained" or "outlined".
- Do we need a design for where to place the icons for buttons with icons?
  - Also, it would be great to have design guidance on when to use icons vs. when not to use icons. 
- I also think we need to specify the case of the text within the button. I agree with @ghislaineguerin that sentence case seems more readable than uppercase.

**Comment by Dhruvi16** [comment]
This is the new iteration of the button component and I have tried to incorporate your suggestions but I do have certain doubts and I have listed them below.
[Link to the doc](https://www.figma.com/file/xYEhAhzGZCwTKmUMqstt2Y/Components?node-id=197%3A2)


Doubts - 
1. I am not sure about the usecases of each size variant and I am not sure if the documentation of the size varaitions is appropriate.
2. I am also not sure about the events associated button I have added four types but I am not sure of their usecases either.
3. I am also not sure about the icon variants and their usecases.

Would love to know your opinion on these points @ghislaineguerin

**Comment by ghislaineguerin** [comment]
@Dhruvi16 Thanks for the update and sorry for the delay in replying.

> I am not sure about the use cases of each size variant and I am not sure if the documentation of the size variations is appropriate.

So far, in the design, we have 3 size variations. The normal or regular used for actions, like saving changes. Then the smaller toolbar icons (see example below):
![image](https://user-images.githubusercontent.com/845767/128508968-6ba5e25c-1440-42ba-9159-4e65df9c4605.png)
And finally, the large ones used for empty states
![image](https://user-images.githubusercontent.com/845767/128508907-f388b8a5-ead1-470c-bbd8-1985a3ae2d0c.png)

> I am also not sure about the events associated button I have added four types but I am not sure of their use cases either.

The only special button we are using now is the 'danger' variation for deleting objects
![image](https://user-images.githubusercontent.com/845767/128509216-125b3181-ff59-4bfc-a089-711a91714de9.png)

> I am also not sure about the icon variants and their use cases.

The icon variants we are using are:
Single icon / no text label
![image](https://user-images.githubusercontent.com/845767/128509289-22a6f03e-fa64-4b5a-8725-66fec2c897f8.png)

Left-aligned icon / text label
![image](https://user-images.githubusercontent.com/845767/128509372-6490ceea-48e3-4589-a2f4-e3f304386dee.png)
